confdgnmi/src/confd_gnmi_server.py

```diff
@@ -153,38 +183,31 @@ def Subscribe(self, request_iterator, context):
         log.info(
             "==> request_iterator=%s context=%s", request_iterator, context)
 
-        def read_sub_request(request_iterator, handler):
-            for req in request_iterator:
-                # TODO check req mode is POLL
-                if hasattr(req, "poll"):
-                    handler.poll()
-                elif hasattr(req, "alias"):
-                    # TODO exception, "alias: request not yet implemented
-                    assert (False)
-                else:
-                    # TODO exception, not expected other type of request
-                    assert (False)
-            handler.stop()
+        def subscribe_rpc_done():
+            log.info("==>")
+            if not handler.is_once():
+                handler.stop()
+            log.info("<==")
 
+        request = next(request_iterator)
         adapter = self.get_connected_adapter(context)
-        handler = adapter.get_subscription_handler()
+        context.add_callback(subscribe_rpc_done)
         # first request, should contain subscription list (`subscribe`)
-        request = next(request_iterator)
-        # TODO exception
         assert hasattr(request, "subscribe")
-        handler.add_subscription_list(request.subscribe)
+        handler = adapter.get_subscription_handler(request.subscribe)
+
         thr = None
-        streaming = (
-                            request.subscribe.mode == gnmi_pb2.SubscriptionList.POLL) or (
-                            request.subscribe.mode == gnmi_pb2.SubscriptionList.STREAM)
-        if streaming:
-            thr = threading.Thread(target=read_sub_request,
-                                   args=(request_iterator, handler,))
+        if not handler.is_once():
+            thr = threading.Thread(target=ConfDgNMIServicer._read_sub_request,
+                                   args=(request_iterator, handler,
+                                         handler.is_poll()))
             thr.start()
         # `yield from` can be used, but to allow altering (e.g. path conversion)
         # response later on we use `for`
-        for response in handler.read(streaming=streaming):
+        for response in handler.read():
+            log.debug("response received")
             yield response
+            log.debug("response yield")
```

Comment:
this will trigger on another call to this generator, so log message might not be appropriate in this form here - may be "delayed" if no "next" iteration is done...

Reply:
updated